docs/gguf.md

```diff
@@ -0,0 +1,372 @@
+    uint8_t _padding[ALIGNMENT - (sizeof(header) % ALIGNMENT)];
+
+    // Tensor infos, which can be used to locate the tensor data.
+    gguf_tensor_info_t tensor_infos[header.tensor_count];
+
+    // Tensor data.
+    //
+    // This is arbitrary binary data corresponding to the weights of the model. This data should be close
+    // or identical to the data in the original model file, but may be different due to quantization or
+    // other optimizations for inference. Any such deviations should be recorded in the metadata or as
+    // part of the architecture definition.
+    //
+    // Each tensor's data must be stored within this array, and located through its `tensor_infos` entry.
+    // The offset of each tensor's data must be a multiple of `ALIGNMENT`, and the space between tensors
+    // should be padded to `ALIGNMENT` bytes.
+    uint8_t tensor_data[];
+};
+```
+
+## Standardized key-value pairs
+
+The following key-value pairs are standardized. This list may grow in the future as more use cases are discovered. Where possible, names are shared with the original model definitions to make it easier to map between the two.
+
+Not all of these are required, but they are all recommended. Keys that are required are bolded. For omitted pairs, the reader should assume that the value is unknown and either default or error as appropriate.
+
+### General
+
+- **`general.architecture: string`**: describes what architecture this model implements. All lowercase ASCII, with only `[a-z0-9]+` characters allowed. Known values include:
+  - `llama`
+  - `mpt`
+  - `gptneox`
+  - `gptj`
+  - `gpt2`
+  - `bloom`
+  - `falcon`
+  - `rwkv`
+- **`general.quantization_version: u32`**: version of quantization scheme
+- `general.file_type: string`: type of the majority of the tensors in the file. This shouldn't have any semantic meaning and should be purely informational, hence the use of `string`.
+- `general.license: string`: SPDX license of the model
+- `general.description: string`: information about the model, including provenance
+- `general.url.original_source: string`: path to the original model that this GGML file was created from
+
+### LLM
+
+In the following, `[llm]` is used to fill in for the name of a specific LLM architecture. They will be used in each architecture's section.
+
+- `[llm].context_length: u32`: size of the maximum supported context
+- `[llm].hidden_size: u32`: embedding layer size
+- `[llm].num_layers: u32`: number of layers
+- `[llm].num_rotary: u32`: `int(hparams["rotary_pct"]*(hparams["hidden_size"]//hparams["num_attention_heads"]))`
+- `[llm].use_parallel_residual: bool`: whether or not the parallel residual logic should be used
+- `[llm].max_seq_len: u32`: Maximum sequence length
+- `[llm].attention.num_heads: u32`: number of attention heads
+- `[llm].attention.alibi_bias_max: f32`: The maximum bias to use for ALiBI
+- `[llm].attention.clip_kqv: f32`: **TODO**: what is this?
```

Comment:
This is a floating point value `C` used to clamp the values of Q, K, V in the range [-C, C]

Reply:
can we call this clamp? clip is kind of loaded...

Reply:
Added this, but I think `kqv_clamp` might be a better name for the same reason.

Reply:
Regarding the naming of things, I have a few opinions that I don't feel super strongly about, so feel free to ignore:

1.
`max_seq_len`
`attention.alibi_bias_max`

This is inconsistent. It should be either `seq_len_max` and `attention.alibi_bias_max` or `max_seq_len` and `attention.max_alibi_bias`

2.
How to decide between `kqv_clamp` and `clamp_kqv`?
The way I usually solve this dilemma is to think what would be more likely to generalize in this context. For example, if I expect to see clamp coefficients for other things in the future then I will go with `clamp_kqv` so that later we can have `clamp_soft_max`, `clamp_norm`, etc., and so the prefix will fit neatly across nearby parameters. On the other hand, if I expect to have more `kqv` related properties, then I would go with `kqv_clamp` so that I can add later `kqv_scale`, `kqv_offset`, etc. Obviously in some cases you cannot tell which of the 2 cases is better, so then you pick one of them at random. For this specific case, I believe that the likelihood of having more `clamp` values in the future is higher, so maybe go with `clamp_kqv`

3.
In `ggml` we are already using extensively `n_head`, `n_past`, n_layer` compared to `num_head`, `num_past` and `num_layer`, so maybe we can do this in the spec too

Reply:
1. Good point, will do `max_alibi_bias`
2. Fine by me, will do
3. `num_` came from the hyperparameters of the GPT-NeoX models:
https://github.com/ggerganov/ggml/blob/23eda2e5c491714daa60d015fb3de8d8f446b88a/examples/gpt-neox/convert-h5-to-ggml.py#L52-L53
but there doesn't seem to be any consistency here, so I'll switch back to `n_` unless I see anything that would conflict with that